Mathlib/RingTheory/Coalgebra/Hom.lean

```diff
@@ -0,0 +1,310 @@
+  [CoalgHomClass F R A B]
+
+/-- Turn an element of a type `F` satisfying `CoalgHomClass F R A B` into an actual
+`CoalgHom`. This is declared as the default coercion from `F` to `A →ₗc[R] B`. -/
+@[coe]
+def toCoalgHom (f : F) : A →ₗc[R] B :=
+  { (f : A →ₗ[R] B) with
+    toFun := f
+    counit_comp := CoalgHomClass.counit_comp f
+    map_comp_comul := CoalgHomClass.map_comp_comul f }
+
+instance instCoeToCoalgHom : CoeHead F (A →ₗc[R] B) :=
+  ⟨CoalgHomClass.toCoalgHom⟩
+
+@[simp]
+theorem counit_comp_apply (f : F) (x : A) : counit (f x) = counit (R := R) x :=
+  LinearMap.congr_fun (counit_comp f) _
+
+@[simp]
+theorem map_comp_comul_apply (f : F) (x : A) :
+    TensorProduct.map f f (comul x) = comul (R := R) (f x) :=
+  LinearMap.congr_fun (map_comp_comul f) _
+
+end CoalgHomClass
+
+namespace CoalgHom
+
+variable {R A B C D : Type*}
+
+section
+
+variable [CommSemiring R] [AddCommMonoid A] [Module R A] [AddCommMonoid B] [Module R B]
+  [AddCommMonoid C] [Module R C] [AddCommMonoid D] [Module R D]
+  [CoalgebraStruct R A] [CoalgebraStruct R B] [CoalgebraStruct R C] [CoalgebraStruct R D]
+
+instance funLike : FunLike (A →ₗc[R] B) A B where
+  coe f := f.toFun
+  coe_injective' f g h := by
+    rcases f with ⟨⟨⟨_, _⟩, _⟩, _, _⟩
+    rcases g with ⟨⟨⟨_, _⟩, _⟩, _, _⟩
+    congr
+
+instance coalgHomClass : CoalgHomClass (A →ₗc[R] B) R A B where
+  map_add := fun f => f.map_add'
+  map_smulₛₗ := fun f => f.map_smul'
+  counit_comp := fun f => f.counit_comp
+  map_comp_comul := fun f => f.map_comp_comul
+
+/-- See Note [custom simps projection] -/
+def Simps.apply {R α β : Type*} [CommSemiring R]
+    [AddCommMonoid α] [Module R α] [AddCommMonoid β]
+    [Module R β] [CoalgebraStruct R α] [CoalgebraStruct R β]
+    (f : α →ₗc[R] β) : α → β := f
+
+initialize_simps_projections CoalgHom (toFun → apply)
+
+@[simp]
+protected theorem coe_coe {F : Type*} [FunLike F A B] [CoalgHomClass F R A B] (f : F) :
+    ⇑(f : A →ₗc[R] B) = f :=
+  rfl
+
+-- removed @[simp], linter complains
+theorem toFun_eq_coe (f : A →ₗc[R] B) : f.toFun = f := rfl
+
+-- are the next 2 declarations necessary? Can already coerce to an `AddMonoidHom`
+/-- The `AddMonoidHom` underlying a coalgebra homomorphism. -/
+@[coe]
+def toAddMonoidHom' (f : A →ₗc[R] B) : A →+ B := (f : A →ₗ[R] B)
+
+instance coeOutAddMonoidHom : CoeOut (A →ₗc[R] B) (A →+ B) :=
+  ⟨CoalgHom.toAddMonoidHom'⟩
+
+@[simp]
+theorem coe_mk {f : A →ₗ[R] B} (h h₁) : ((⟨f, h, h₁⟩ : A →ₗc[R] B) : A → B) = f :=
+  rfl
+
+@[norm_cast]
+theorem coe_mks {f : A → B} (h₁ h₂ h₃ h₄) : ⇑(⟨⟨⟨f, h₁⟩, h₂⟩, h₃, h₄⟩ : A →ₗc[R] B) = f :=
+  rfl
+
+@[simp, norm_cast]
+theorem coe_linearMap_mk {f : A →ₗ[R] B} (h h₁) : ((⟨f, h, h₁⟩ : A →ₗc[R] B) : A →ₗ[R] B) = f :=
+  rfl
+
+@[simp]
+theorem toLinearMap_eq_coe (f : A →ₗc[R] B) : f.toLinearMap = f :=
+  rfl
+
+@[simp, norm_cast]
+theorem coe_toLinearMap (f : A →ₗc[R] B) : ⇑(f : A →ₗ[R] B) = f :=
+  rfl
+
+-- simp can prove this
```

Comment:
Can it? Then why isn't there a lint error?

Reply:
Yep. I don't know why there isn't a lint error. I'll remove the `@[simp]` anyway.